lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -3338,9 +3338,13 @@ bool TypeChecker::convertToType(Expr *&expr, Type type, DeclContext *dc,
 
   // Attempt to solve the constraint system.
   SmallVector<Solution, 4> viable;
-  if ((cs.solve(expr, viable) || viable.size() != 1) &&
-      cs.salvage(viable, expr)) {
-    return true;
+  if (cs.solve(expr, viable) || viable.size() != 1) {
+    // If the cs don't have a type for this expression we don't need to attemp salvage.
+    if (!cs.hasType(expr) && expr->getType()->is<UnresolvedType>())
```

Comment:
Can you try to get to the root cause of why this type remains unresolved?  It’s not sound to ignore this.

Reply:
You can use `-Xfrontend -debug-constraints` to get detailed information from
the solver. But it’s often easier to set breakpoints in CSGen and see how we’re setting up this constraint system. A lot of these crashes are from bogus information entering the solver from beforehand.

Reply:
There's some dodgy code in CSApply where we set up child constraint systems when applying a solution involving a cast; if that fails, we 'return null' from applying the solution which leaves us which a half-checked AST. Could be that?

Reply:
Hey guys, thank's for the review :)
@slavapestov Do you mean [this](https://github.com/apple/swift/blob/fb8bd3a056c4958a47bfc67d2a546d51fc53d4c5/lib/Sema/CSApply.cpp#L3711) return null? I set a breakpoint and it actually doesn't fall there in this case. But actually, if we try `cs.hasType(sub)` using the cs instance here before call `tc.convertToType..` it returns `true`, but calling it with a new cs inside the `convertToType` method gives us `false`. This new cs inside is being created with the same tc ref of the previous. So I'm trying to look what's different in the two, as @CodaFi mention, to find the root cause why this remains unsolved on this new cs.

Reply:
Forget the previous comment, don't make much sense, I'm just learning about these things as it goes 😅, @CodaFi I found it is falling on solving the type`cs.addConstraint(ConstraintKind::Conversion, expr->getType(), type, cs.getConstraintLocator(expr));` [here](https://github.com/apple/swift/blob/fb8bd3a056c4958a47bfc67d2a546d51fc53d4c5/lib/Sema/TypeCheckConstraints.cpp#L3328), not sure why yet, still looking. Maybe because of what you mention @slavapestov?